(close #1704) replace_with takes into account named arguments

sergisiso opened the pull request.

sergisiso commented:
This is ready for review. Probably for @rupertford . I took your suggestion from the issue discussion of modifying repalce_with but since its the parent type that matters I have not specialize it but used a `hasattr(self.parent,` instead.

sergisiso commented:
@rupertford I made an extra commit to fix a small bug that I found with the named_args. Now I ran it with the full Nemo codebase. It is still ready for review.

rupertford commented on doc/user_guide/psyir.rst:
```diff
@@ -726,9 +726,11 @@ methods.
+argument is replaced by removing and inserting a new node then it is
+assumed that this argument is no longer a named argument. If it is
+replaced with the `replace_with` method, it has a `keep_name_in_context`
+argument to chose the desired behaviour (defaults to True).
```
chose -> choose

sergisiso replied:
Fixed

rupertford commented on src/psyclone/psyir/nodes/call.py:
```diff
@@ -233,7 +232,7 @@ def replace_named_arg(self, existing_name, arg):
                 f"in the existing arguments.")
         self.children[index] = arg
-        self._argument_names[index] = (id(arg), named_arg[1])
+        self._argument_names[index] = (id(arg), existing_name)
```
This method no longer does what it says. It used to  replace one named argument with another. Now it replaces a named argument with an argument and keeps the original name irrespective of whether the replacing argument is named or not. Could you please reflect this change in the description. Is it also an error if the new argument is a named argument as we just throw the name away????

sergisiso replied:
I don't think that's what it did. We just provided a new node also before, never a new name. I just changed name_arg[1] to existing_name because they already matched in the for loop and named_arg was producing a pylint error for being used outside the loop were it is defined.

> Is it also an error if the new argument is a named argument as we just throw the name away????

No, that's the id of the node that is being replaced.

That said, I modified the docstring to make this clear and also updated the equivalent Operation node.

rupertford commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1313,16 +1313,19 @@ def scope(self):
             tree.
         :type node: :py:class:`psyclone.psyir.nodes.node`
+        :param bool keep_name_in_context: whether to conserve the name \
+            referencing this node when it is within a parent with named nodes.
```
I'm not sure it makes sense to mention parent here as that is an implementation? Perhaps just remove (when it is ... named nodes.)

sergisiso replied:
Done

rupertford reviewed (changes requested):
Looks good Sergi. Pleasingly small. I've made a few comments inline. I'm guessing the documentation should be updated somewhere to describe what is going on, possibly in the psyir section of the user or developer guide? Is it also worth adding to the existing psyir examples to demonstrate named args if we don't already do so? Perhaps that is another PR.

sergisiso commented:
@rupertford This is ready for review again. I added documentation and examples as suggested. Also I think there was some confusion about the replace_named_arg method, I updated that docstring as well in both Call and Operation nodes.

rupertford reviewed (approved):
Requested changed have been made. All looks good. Approving.
